source/system/TextDatasource.js

```diff
@@ -2,18 +2,95 @@
 
 const iocane = require("iocane").crypto;
 
-var Archive = require("./Archive.js"),
-    Credentials = require("./credentials.js"),
-    signing = require("../tools/signing.js"),
-    encoding = require("../tools/encoding.js"),
-    createDebug = require("../tools/debug.js"),
-    historyTools = require("../tools/history.js");
-
+const Archive = require("./Archive.js");
+const Credentials = require("./credentials.js");
+const signing = require("../tools/signing.js");
+const encoding = require("../tools/encoding.js");
+const createDebug = require("../tools/debug.js");
+const historyTools = require("../tools/history.js");
 const registerDatasource = require("./DatasourceAdapter.js").registerDatasource;
 
 const debug = createDebug("text-datasource");
 
 /**
+ * Current appointed callback for decrypting archive content
+ * @type {Function}
+ * @private
+ */
+let __appointedEncToHistoryCB = convertEncryptedContentToHistory,
+    /**
+     * Current appointed callback for encrypting archive history
+     * @type {Function}
+     * @private
+     */
+    __appointedHistoryToEncCB = convertHistoryToEncryptedContent;
+
+/**
+ * Convert encrypted text to an array of commands (history)
+ * @param {String} encText The encrypted archive content
+ * @param {Credentials} credentials A credentials instance that has a password, keyfile
+ *  or both
+ * @returns {Promise.<Array>} A promise that resolves with an array of commands
+ */
+function convertEncryptedContentToHistory(encText, credentials) {
+    const { password, keyfile } = processCredentials(credentials);
+    return Promise.resolve(encText)
+        .then(function __stripSignature(data) {
+            if (!signing.hasValidSignature(data)) {
+                throw new Error("No valid signature in archive");
+            }
+            return signing.stripSignature(data);
+        })
+        .then(function __decryptUsingKeyFile(encryptedData) {
+            // optionally decrypt using a key file
+            return keyfile ?
+                iocane.decryptWithKeyFile(encryptedData, keyfile) :
+                encryptedData;
+        })
+        .then(function __decryptUsingPassword(encryptedData) {
+            // optionally decrypt using a password
+            return password ?
+                iocane.decryptWithPassword(encryptedData, password) :
+                encryptedData;
+        })
+        .then(function __marshallHistoryToArray(decrypted) {
+            if (decrypted && decrypted.length > 0) {
```

Comment:
What if both password and keyfile are absent, and this `decrypted` variable still is the encrypted data?